myhoard/web_server.py

```diff
@@ -78,6 +79,24 @@ async def backup_list(self, _request):
                     response["backups"] = self.controller.state["backups"]
                 return json_response(response)
 
+    async def backup_preserve(self, request):
+        with self._handle_request(name="backup_preserve"):
+            stream_id = request.match_info["stream_id"]
+            body = await self._get_request_json(request)
+            preserve_until = body.get("preserve_until")
+
+            if preserve_until is not None:
+                try:
+                    preserve_until = datetime.fromisoformat(preserve_until).astimezone(timezone.utc)
+                    now = datetime.now(timezone.utc)
+                    if preserve_until < now:
+                        raise BadRequest("`preserve_until` must be a date in the future.")
+                except ValueError:
+                    raise BadRequest("`preserve_until` must be a valid isoformat datetime string.")
+
+            self.controller.mark_backup_preservation(stream_id=stream_id, preserve_until=preserve_until)
```

Comment:
Maybe some misunderstanding here but this is still a blocking call that makes blocking network IO and we must not do that from async context as the API will be blocked. We should instead update local state to indicate that backup is expected to be preserved and then use async wait here to wait for the request to actually go through.

Reply:
hah that's true... maybe having a queue for backup preservation requests and later make the controller handle it?  Because the request might take a bit of time (depending how long it takes refreshing backups), so maybe handling the batch of requests and later refreshing might be ideal. I just find a bit misleading returning "success"

Reply:
Just having some `{"pending_preservation_requests": {"<stream_id_1>": "<timestamp_1>"|null, "<stream_id_2>": "<timestamp_2>"|null}}` would presumably be quite sufficient way of storing this information in controller state. It could then drop those entries from `pending_preservation_requests` once the data has actually been updated.

I think the `while True: await asyncio.sleep(1); check_if_operation_is_done_yet` pattern used in some other handler should be OK for this. If the operation takes a very long time, like over 60 seconds, this could raise timeout error to avoid returning misleading response.

Reply:
well... maybe not even a queue, just updating the state is enough

Reply:
:+1:  alright, will add `wait_for_applied_preservation` as optional in the request's body and wait only in this case (we do something similar for backup_create)